base/unicode/utf8.jl

```diff
@@ -101,26 +101,33 @@ sizeof(s::UTF8String) = sizeof(s.data)
 
 lastidx(s::UTF8String) = length(s.data)
 
+isfirstbyte(c::UInt8) = (c & 0xc0) != 0x80 # == !is_valid_continuation(c)
+
 isvalid(s::UTF8String, i::Integer) =
-    (1 <= i <= endof(s.data)) && !is_valid_continuation(s.data[i])
+    (1 <= i <= endof(s.data)) && isfirstbyte(s.data[i])
 
 const empty_utf8 = UTF8String(UInt8[])
-
 function getindex(s::UTF8String, r::UnitRange{Int})
     isempty(r) && return empty_utf8
     i, j = first(r), last(r)
     d = s.data
-    if i < 1 || i > length(s.data)
+    if i < 1 || i > length(d)
         throw(BoundsError(s, i))
     end
     if is_valid_continuation(d[i])
         throw(UnicodeError(UTF_ERR_INVALID_INDEX, i, d[i]))
     end
-    if j > length(d)
-        throw(BoundsError())
+    if j < 1 || j > length(d)
+        throw(BoundsError(s, i))
+    end
+    #ensure j is the first or last byte of a character
```

Comment:
See https://github.com/JuliaLang/julia/issues/14158#issuecomment-161050256, let's continue the discussion there.